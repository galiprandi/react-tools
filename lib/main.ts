export { Input } from "./components/Input";
export { Label } from "./components/Label";
