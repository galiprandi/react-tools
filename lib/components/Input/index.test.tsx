import { describe, it, expect, afterEach, vi } from 'vitest'
import { cleanup, fireEvent, render } from '@testing-library/react'

import { Input } from '../../main.ts'

describe('<Input />', () => {
    afterEach(cleanup)

    it('should be render a input', () => {
        const testId = 'test-1'
        const { getByTestId } = render(
            <Input id={testId} type="text" name="name" data-testid={testId} />,
        )

        expect(getByTestId(testId)).toBeDefined()
    })

    it('should be render a label', () => {
        const testId = 'test-2'
        const { getByTestId, getByText } = render(
            <Input
                id={testId}
                type="text"
                name="name"
                label="Enter your name:"
                data-testid={testId}
            />,
        )

        expect(getByTestId(testId)).toBeDefined()
        expect(getByText(/enter your name/i)).toBeDefined()
    })

    it('should be render a datalist with options', () => {
        const testId = 'test-3'
        const datalist = ['John Doe', 'Jane Doe', 'John Smith']
        const { getByTestId, container } = render(
            <Input
                type="text"
                name="name"
                label="Enter your name:"
                data-testid={testId}
                datalist={datalist}
            />,
        )

        expect(getByTestId(testId)).toBeDefined()
        expect(container.querySelector('datalist')).toBeDefined()
        expect(container.querySelectorAll('option')).toHaveLength(
            datalist.length,
        )
        expect(container).toMatchSnapshot()
    })

    it('should be call onChange, onChangeValue & onChangeDebounce', () => {
        const testId = 'test-4'
        const onChange = vi.fn()
        const onChangeValue = vi.fn()
        const onChangeDebounce = vi.fn()

        const { getByTestId } = render(
            <Input
                id={testId}
                type="text"
                name="name"
                label="Enter your name:"
                data-testid={testId}
                onChange={onChange}
                onChangeValue={onChangeValue}
                onChangeDebounce={onChangeDebounce}
            />,
        )

        // Type 'test value' in input
        const input = getByTestId(testId)
        fireEvent.change(input, { target: { value: 'test value' } })

        expect(onChange).toHaveBeenCalled()
        expect(onChangeValue).toHaveBeenCalledWith('test value')
        expect(onChangeDebounce).toHaveBeenCalledWith('test value')
    })

    it('should be transform value to uppercase', () => {
        const testId = 'test-5'
        const onChangeValue = vi.fn()
        const transform = 'toUpperCase'

        const { getByTestId } = render(
            <Input
                id={testId}
                type="text"
                name="name"
                defaultValue={'test value'}
                data-testid={testId}
                transform="toUpperCase"
                onChangeValue={onChangeValue}
            />,
        )

        // Type 'test value' in input
        const input = getByTestId(testId)
        fireEvent.change(input, { target: { value: 'test value' } })

        expect(onChangeValue).toHaveBeenCalledWith('TEST VALUE')
    })

    it('should be called transformFn', () => {
        const testId = 'test-5'
        const onChangeValue = vi.fn()
        const transform = 'toUpperCase'

        const { getByTestId } = render(
            <Input
                id={testId}
                type="text"
                name="name"
                data-testid={testId}
                transformFn={(value) => `${value}.transformed`}
                onChangeValue={onChangeValue}
            />,
        )

        // Type 'test value' in input
        const input = getByTestId(testId)
        fireEvent.change(input, { target: { value: 'value' } })

        expect(onChangeValue).toHaveBeenCalledWith('value.transformed')
    })
})
