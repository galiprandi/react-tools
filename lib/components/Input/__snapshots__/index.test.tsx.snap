// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Input /> > should be render a datalist with options 1`] = `
<div>
  <label>
    Enter your name:
    <input
      data-testid="test-3"
      list=":r0:-datalist"
      name="name"
      type="text"
      value=""
    />
    <datalist
      id=":r0:-datalist"
    >
      <option
        value="John Doe"
      />
      <option
        value="Jane Doe"
      />
      <option
        value="John Smith"
      />
    </datalist>
  </label>
</div>
`;
